### PR TITLE
fixed the Khan Academy php scrapper refs #778

### PR DESCRIPTION
**Pull request by ajakov**
Hi everybody,

I fixed the PHP version. 

They are using a combination of a data that can be found in the HTML and a timestamp in milliseconds to create subsequent api request that contains the data. So I copied the request from the browser to postman and from postman exported to php code and generated the url in the described way.

### CODE REVIEW

File: scrapers/php/khan.php [hunk 1]
```diff
@@ -1,13 +1,98 @@
...
+ * @file
+ * This file can be used to fetch the energy points
+ * from Khan Academy user profile
+ * usage: php khan.php khanAcademyUsername
+ *
+ * @category CLI_Script
+ * @author   "Aleksandar Jakovljevic aleksandarjakovljevic@gmail.com"
+ * @license  https://www.gnu.org/licenses/gpl-3.0.html GNU GENERAL PUBLIC LICENSE
+ * @link     https://github.com/slurpcode/slurp
+ * @Package  scrappers
```
Comment (jbampton) [inline scrapers/php/khan.php:17]:
```suggestion
 * @Package  scrapers
```

File: scrapers/php/khan.php [hunk 2]
```diff
@@ -1,13 +1,96 @@
...
...{\\n      name\\n      imageSrc\\n      __typename\\n    }\\n    soundOn\\n    muteVideos\\n    prefersReducedMotion\\n    noColorInVideos\\n    autocontinueOn\\n    avatar {\\n      name\\n      imageSrc\\n      __typename\\n    }\\n    hasChangedAvatar\\n    newNotificationCount\\n    canHellban: hasPermission(name: \\"can_ban_users\\", scope: GLOBAL)\\n    canMessageUsers: hasPermission(name: \\"can_send_moderator_messages\\", scope: GLOBAL)\\n    discussionBanned\\n    isSelf: isActor\\n    hasStudents: hasCoachees\\n    hasClasses\\n    hasChildren\\n    hasCoach\\n    badgeCounts\\n    homepageUrl\\n    isMidsignupPhantom\\n    includesDistrictOwnedData\\n    preferredKaLocale {\\n      id\\n      kaLocale\\n      status\\n      __typename\\n    }\\n    underAgeGate {\\n      parentEmail\\n      daysUntilCutoff\\n      approvalGivenAt\\n      __typename\\n    }\\n    authEmails\\n    signupDataIfUnverified {\\n      email\\n      emailBounced\\n      __typename\\n    }\\n    pendingEmailVerifications {\\n      email\\n      unverifiedAuthEmailToken\\n      __typename\\n    }\\n    tosAccepted\\n    shouldShowAgeCheck\\n    __typename\\n  }\\n  actorIsImpersonatingUser\\n}\\n"
```
[the remark this answers is not in the document]
Comment (radoy) [reply scrapers/php/khan.php:53]: For the CURLOPT_POSTFIELDS value, store it in a variable to keep clean code.

File: scrapers/php/khan.php [hunk 3]
```diff
@@ -1,13 +1,96 @@
...
+
+    // Note: Using a string here to prevent loss of precision
+    // in case of "overflow" (PHP converts it to a double)
+    return sprintf('%d%03d', $comps[1], $comps[0] * 1000);
+}
+
+function getFullUserProfile($dashParameter, $user)
+{
+
+    $url = "https://www.khanacademy.org/api/internal/graphql/getFullUserProfile?lang=en&_=" . $dashParameter;
```
[the remark this answers is not in the document]
Comment (radoy) [reply scrapers/php/khan.php:35]: The 'https://www.khanacademy.org/', defines it as a constant.

File: scrapers/php/khan.php [hunk 4]
```diff
@@ -1,13 +1,96 @@
...
+            'user-agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.4430.212 Safari/537.36',
+            'content-type: application/json',
+            'origin: https://www.khanacademy.org',
+            'sec-fetch-site: same-origin',
+            'sec-fetch-mode: cors',
+            'sec-fetch-dest: empty',
+            'referer: https://www.khanacademy.org/profile/' . $user . '/',
+            'accept-language: en-GB,en;q=0.9,sr-RS;q=0.8,sr;q=0.7,en-US;q=0.6'
+        ),
+    ));
```
[the remark this answers is not in the document]
Comment (radoy) [reply scrapers/php/khan.php:71]: The 'https://www.khanacademy.org/', defines it as a constant.

File: scrapers/php/khan.php [hunk 5]
```diff
@@ -1,13 +1,96 @@
...
+    $content = get_content($url);
+    preg_match('/\"x-ka-static-version\":\"([^\"]*)\"/m', $content, $matches);
+    $static_version = $matches[1];
+    $millisecondsTime = millitime();
+    $dashParameter = $static_version . '_' . $millisecondsTime;
+
+    $fullUserProfile = getFullUserProfile($dashParameter, $user);
+    $fullUserProfileDecoded = json_decode($fullUserProfile, true);
+
+    return isset($fullUserProfileDecoded['data']['user']['points']) ? $fullUserProfileDecoded['data']['user']['points'] : '';
```
[the remark this answers is not in the document]
Comment (radoy) [reply scrapers/php/khan.php:91]: The 'https://www.khanacademy.org/', defines it as a constant.

[the remark this answers is not in the document]
Comment (radoy) [reply scrapers/php/khan.php:91]: $static_version change it to $staticVersion

**Review by radoy — COMMENTED** [review]
Would it be nice if you define the "https://www.khanacademy.org" as a constant. If you need to change the domain name, that's should be just in one place.

**Comment by ajakov** [comment]
> Would it be nice if you define the "https://www.khanacademy.org" as a constant. If you need to change the domain name, that's should be just in one place.

@radoy actually, it would never change, since it is a scrapper made specifically for Khan Academy.

File: scrapers/php/khan.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (ajakov) [reply scrapers/php/khan.php:53]: Why? It is not a variable. It is a very specific case, happening only once and used only once. If we were making a general case class that would make sense. Like this, it's a waste of time.

**Comment by BaseMax** [comment]
Hey you guys;
What happened for this PR? There is problems something's not solved yet? Please make clear status of this.

**Comment by ajakov** [comment]
> Hey you guys;
> What happened for this PR? There is problems something's not solved yet? Please make clear status of this.

me and @radoy had some disagreement about whether some things should be variables/constants or hardcoded. :)
Other then that, script is working.

**Comment by ajakov** [comment]
IMO, it doesn't make sense to move it into variables because it is a highly specific case, and would not work for any other values. If you change the domain - it wouldn't work. If you change the headers - it won't work.
What makes sense, however, is to actually add functions for scrapping different information. 
How it works: 

1. Initial HTML is fetched
2. Page makes subsequent requests to fetch the JSON data from the API. The requests need to be "signed" using the information found in HTML ($staticVersion) and current timestamp in miliseconds. 

Now, full user data is visible, and we could add function to save/show other relevant info. 
Also, there are other calls that can be made in the same manner. 

Moving the mentioned pieces of code into variables would not add any real value to the script. But if you guys all say that I should move it, I will move it, of course.

**Comment by BaseMax** [comment]
Thank you Aleksandar and Muhammad for your participation in this PR.

Best wishes.

**Pull request closed without merging**